### PR TITLE
hopper: 4.5.29 -> 5.5.3; libffi_3_3: init

### PR DESCRIPTION
**Pull request by armeenm**
###### Description of changes
Hopper 5.5.3 requires libffi.so.7 which is provided by libffi 3.3. This PR inits a package libffi7 akin to [Ubuntu's](https://packages.ubuntu.com/focal/libffi7) for the purpose of bumping Hopper to 5.5.3 as 4.5.29 relies on python2 (#148779).

###### Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [x] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [x] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [22.05 Release Notes (or backporting 21.11 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2205-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
  - [ ] (Release notes changes) Ran `nixos/doc/manual/md-to-db.sh` to update generated release notes
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

**Comment by Mindavi** [comment]
The naming of libffi7 is kinda weird, I find.

**Comment by armeenm** [comment]
Yeah I wasn't really sure what to name it so I just went with precedent, but let me know if you have any better ideas.

File: pkgs/development/libraries/libffi/7.nix [hunk 1]
```diff
@@ -0,0 +1,69 @@
+{ lib, stdenv, fetchurl, fetchpatch
+, autoreconfHook
+
+, doCheck ? true # test suite depends on dejagnu which cannot be used during bootstrapping
+, dejagnu
+}:
+
+# Note: this package is used for bootstrapping fetchurl, and thus
```
Comment (Mindavi) [inline pkgs/development/libraries/libffi/7.nix:8]: This comment not true for this alternative version of the libffi package.

File: pkgs/top-level/all-packages.nix [hunk 2]
```diff
@@ -18482,6 +18482,7 @@ with pkgs;
   libffcall = callPackage ../development/libraries/libffcall { };
 
   libffi = callPackage ../development/libraries/libffi { };
+  libffi7 = callPackage ../development/libraries/libffi/7.nix { };
```
Comment (Mindavi) [inline pkgs/top-level/all-packages.nix:18485]: maybe just call it libffi3_3?

Comment (Mindavi) [reply pkgs/top-level/all-packages.nix:18485]: or libffi_so7

File: pkgs/development/libraries/libffi/7.nix [hunk 3]
```diff
@@ -0,0 +1,69 @@
...
+      interface is the popular name for the interface that allows code
+      written in one language to call code written in another
+      language.  The libffi library really only provides the lowest,
+      machine dependent layer of a fully featured foreign function
+      interface.  A layer must exist above libffi that handles type
+      conversions for values passed between the two languages.
+    '';
+    homepage = "http://sourceware.org/libffi/";
+    license = licenses.mit;
+    maintainers = with maintainers; [ matthewbauer ];
```
Comment (Mindavi) [inline pkgs/development/libraries/libffi/7.nix:66]: cc @matthewbauer are you willing to maintain this?

**Comment by Mindavi** [comment]
Added some comments to improve this, in general this LGTM. But the naming will throw off some people, so I think it's better to improve it somewhat, unless others disagree on that.

File: pkgs/top-level/all-packages.nix (hunk 2 above)
Comment (armeenm) [reply pkgs/top-level/all-packages.nix:18485]: I'll go with libffi_3_3 to match most other things in nixpkgs (e.g. terraform_0_15).

Comment (Mindavi) [reply pkgs/top-level/all-packages.nix:18485]: SGTM

File: pkgs/development/tools/analysis/hopper/default.nix [hunk 4]
```diff
@@ -66,6 +64,7 @@ stdenv.mkDerivation rec {
     substituteInPlace "$out/share/applications/hopper-${rev}.desktop" \
       --replace "Exec=/opt/hopper-${rev}/bin/Hopper" "Exec=$out/bin/hopper"
   '';
+    #ln -s ${lib.getLib libffi}/lib/libffi.so $out/lib/libffi.so.6
```
Comment (Mindavi) [inline pkgs/development/tools/analysis/hopper/default.nix:67]: Stray comment?

File: pkgs/development/tools/analysis/hopper/default.nix [hunk 5]
```diff
@@ -55,7 +54,6 @@ stdenv.mkDerivation rec {
       $out/lib
 
     # we already ship libffi.so.7
```
Comment (Mindavi) [inline pkgs/development/tools/analysis/hopper/default.nix:56]: Should this comment stay?

File: pkgs/development/libraries/libffi/3.3.nix [hunk 6]
```diff
@@ -0,0 +1,64 @@
...
+      interface is the popular name for the interface that allows code
+      written in one language to call code written in another
+      language.  The libffi library really only provides the lowest,
+      machine dependent layer of a fully featured foreign function
+      interface.  A layer must exist above libffi that handles type
+      conversions for values passed between the two languages.
+    '';
+    homepage = "http://sourceware.org/libffi/";
+    license = licenses.mit;
+    maintainers = with maintainers; [ matthewbauer ];
```
Comment (Mindavi) [inline pkgs/development/libraries/libffi/3.3.nix:61]: I propose to remove matthewbauer unless he agrees to maintain this.

**Review by Mindavi — COMMENTED** [review]
Some small comments, overall LGTM

File: pkgs/development/libraries/libffi/3.3.nix (hunk 6 above)
Comment (armeenm) [reply pkgs/development/libraries/libffi/3.3.nix]: Replaced with myself.

**Comment by Mindavi** [comment]
Thabks for sticking through!

**Pull request merged**